### PR TITLE
Spike/multipath

### PR DESCRIPTION
**Pull request by squeedee**
Looking at the work involved for multipath [#570].

I feel like this PR contains the bulk of the work. 

I don't see any significant hurdles to this implementation, but I do wish we had generics already.

### CODE REVIEW

File: pkg/realizer/workload/component.go [hunk 1]
```diff
@@ -70,10 +70,23 @@ func (r *resourceRealizer) Do(ctx context.Context, resource *v1alpha1.SupplyChai
...
+	}
+
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
 	if len(resource.TemplateRef.Options) > 0 {
-		templateName, err = r.findMatchingTemplateName(resource, supplyChainName)
+		matchedOption, err = r.findMatchingTemplateName(resource, supplyChainName, outputs)
+		templateName = matchedOption.Name
```
Comment (cirocosta) [inline pkg/realizer/workload/component.go]: uh, this assignment should happen after the err check

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13fdf53]

goroutine 673 [running]:
github.com/vmware-tanzu/cartographer/pkg/realizer/workload.(*resourceRealizer).Do(0xc000bd8100, {0x1935b58, 0xc000dce5d0}, 0xc0006e2fc0, {0xc0006532a0, 0x1}, 0x1)
        /home/cirocosta/dev/vmware-tanzu/cartographer/pkg/realizer/workload/component.go:89 +0x2b3
github.com/vmware-tanzu/cartographer/pkg/realizer/workload.(*realizer).Realize(0x1720ece, {0x1935b58, 0xc000dce5d0}, {0x190a9a0, 0xc000bd8100}, 0xc000423ba0)
        /home/cirocosta/dev/vmware-tanzu/cartographer/pkg/realizer/workload/realizer.go:53 +0x2b1
github.com/vmware-tanzu/cartographer/pkg/controller/workload.(*Reconciler).Reconcile(0xc0003c4a80, {0x1935b58, 0xc000dce360}, {{{0xc0006adf40, 0x7}, {0xc0006adf36, 0xa}}})
        /home/cirocosta/dev/vmware-tanzu/cartographer/pkg/controller/workload/reconciler.go:124 +0xc63
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0004ce580, {0x1935b58, 0xc000dce2d0}, {{{0xc0006adf40, 0x165d820}, {0xc0006adf36, 0x415594}}})
```

**Comment by cirocosta** [comment]
heey I was playing around with it, but found a little weird behavior - sometimes it seems to select the right DAG path, but sometimes it doesn't, not sure why.

the example I'm working with is as follows:

```
if workload.spec.image:
  image-provider <--[img]--- deployer

if workload.spec.source.git.url:
  source-provider <--[src]--- image-builder <--[img]-- deployer
```

soo, I've got this supply chain resources:

```yaml
  resources:
    - name: source-provider
      templateRef:
        kind: ClusterSourceTemplate
        name: git-repository

    - name: image-builder
      selector:
        matchFields:
          - key: workload.spec.source.git.url
            operator: Exists
      templateRef:
        kind: ClusterImageTemplate
        name: kpack-image
      sources:
        - resource: source-provider
          name: source

    - name: image-provider
      selector:
        matchFields:
          - key: workload.spec.image
            operator: Exists
      templateRef:
        kind: ClusterImageTemplate
        name: image-repository

    - name: deployer
      templateRef:
        kind: ClusterTemplate
        options:
          - name: kappctrl-app
            images:
              - resource: image-builder
                name: image
          - name: kappctrl-app
            images:
              - resource: image-provider
                name: image
```

and the following workloads:

```yaml
---
apiVersion: carto.run/v1alpha1
kind: Workload
metadata:
  name: test-from-source
  labels:
    app.kubernetes.io/part-of: test-from-source
    apps.tanzu.vmware.com/workload-type: web
spec:
  params:
    - name: service_account
      value: test-basic
  serviceAccountName: test-basic
  source:
    git:
      ref:
        branch: main
      url: https://github.com/kontinue/hello-world

---
apiVersion: carto.run/v1alpha1
kind: Workload
metadata:
  name: test-from-image
  labels:
    app.kubernetes.io/part-of: test-from-image
    apps.tanzu.vmware.com/workload-type: web
spec:
  params:
    - name: service_account
      value: test-basic
  serviceAccountName: test-basic
  image: ghcr.io/cirocosta/cartographer@sha256:f721c277ea1eab25a87b938ba162f179c758bbd16d2d7c64b6a9c109f29e3687
```

it's pretty cool to see that we indeed get to the two different trees

```
NAMESPACE  NAME                               READY  REASON                AGE
default    Workload/test-from-image           False  TemplateStampFailure  12m
default    ├─App/test-from-image              -                            11m
default    └─ImageRepository/test-from-image  True   Ready                 12m


vs


NAMESPACE  NAME                                              READY  REASON               AGE
default    Workload/test-from-source                         True   Ready                9m3s
default    ├─App/test-from-source                            -                           7m41s
default    ├─GitRepository/test-from-source                  True   GitOperationSucceed  9m2s
default    └─Image/test-from-source                          True                        8m17s
default      ├─Build/test-from-source-build-1                -                           8m17s
default      │ └─Pod/test-from-source-build-1-build-pod      False  PodCompleted         8m17s
default      ├─PersistentVolumeClaim/test-from-source-cache  -                           8m17s
default      └─SourceResolver/test-from-source-source        True                        8m17s
```

but, e.g., right now we're seeing `TemplateStampFailure` on the `test-from-image` Workload because it thinks it should the the `source -> image-builder -> deployer` path.

do you spot something wrong in that supplychain? any thoughts on what might be causing that?

thx!

**Comment by squeedee** [comment]
@cirocosta You want this:
```yaml
  resources:
    - name: source-provider
      selector:
        matchFields:
          - key: workload.spec.source.git.url
            operator: Exists
      templateRef:
        kind: ClusterSourceTemplate
        name: git-repository

    - name: image-builder
      templateRef:
        kind: ClusterImageTemplate
        name: kpack-image
      sources:
        - resource: source-provider
          name: source

    - name: image-provider
      selector:
        matchFields:
          - key: workload.spec.image
            operator: Exists
      templateRef:
        kind: ClusterImageTemplate
        name: image-repository

    - name: deployer
      templateRef:
        kind: ClusterTemplate
        options:
          - name: kappctrl-app
            images:
              - resource: image-builder
                name: image
          - name: kappctrl-app
            images:
              - resource: image-provider
                name: image
```

**Comment by cirocosta** [comment]
nice!

another little one - I noticed that the supplychain resource list I put above would actually not work when running a setup with the admission webhooks running

e.g., with that one we currently get:


```
kapp: Error: Applying create clustersupplychain/supply-chain (carto.run/v1alpha1) cluster:
  Creating resource clustersupplychain/supply-chain (carto.run/v1alpha1) cluster:
    API server says: admission webhook "supply-chain-validator.cartographer.com" denied the request: error validating clustersupplychain [supply-chain]: duplicate template name [kappctrl-app] found in options for resource [deployer] (reason: error validating clustersupplychain [supply-chain]: duplicate template name [kappctrl-app] found in options for resource [deployer])
```

from https://github.com/vmware-tanzu/cartographer/blob/0cf9ae2eda6e37827c833373f16d679636bf8afd/pkg/apis/v1alpha1/cluster_supply_chain_validations.go#L36-L48

**Pull request closed without merging**